src/unix/bsd/freebsdlike/freebsd/mod.rs

```diff
@@ -1396,28 +1437,62 @@ cfg_if! {
                 self.sigev_notify == other.sigev_notify
                     && self.sigev_signo == other.sigev_signo
                     && self.sigev_value == other.sigev_value
-                    && self.sigev_notify_thread_id
-                        == other.sigev_notify_thread_id
+                    // sigev_notify indicates which union fields are valid
+                    && match self.sigev_notify {
+                        ::SIGEV_NONE => true,
+                        ::SIGEV_SIGNAL => true,
+                        ::SIGEV_THREAD => unsafe {
+                            self._sigev_un._sigev_thread
+                                == other._sigev_un._sigev_thread
+                        },
+                        ::SIGEV_KEVENT => unsafe {
+                            self._sigev_un._kevent_flags
+                                == other._sigev_un._kevent_flags
+                        },
+                        ::SIGEV_THREAD_ID => unsafe {
+                            self._sigev_un._threadid
+                                == other._sigev_un._threadid
+                        },
```

Comment:
All these unsafe blocks are reading and comparing potentially uninitialized data from the union.

```rust
fn main() {
    let sigevent = libc::sigevent {
        sigev_notify: libc::SIGEV_THREAD,
        sigev_signo: 0,
        sigev_value: libc::sigval {
            sival_ptr: 0 as *mut _,
        },
        _sigev_un: libc::__c_anonymous_sigev_un {
            _kevent_flags: 0,
        },
    };
    let _eq = sigevent == sigevent;
}
```

```console
error: Undefined Behavior: type validation failed: encountered uninitialized raw pointer
    --> /git/libc/src/unix/bsd/freebsdlike/freebsd/mod.rs:233:1
     |
233  | / s! {
234  | |     pub struct aiocb {
235  | |         pub aio_fildes: ::c_int,
236  | |         pub aio_offset: ::off_t,
...    |
1013 | |     }
1014 | | }
     | |_^ type validation failed: encountered uninitialized raw pointer
     |
     = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
     = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
             
     = note: inside `<libc::__c_anonymous_sigev_thread as std::cmp::PartialEq>::eq` at /git/libc/src/unix/bsd/freebsdlike/freebsd/mod.rs:876:9
     = note: inside `<libc::sigevent as std::cmp::PartialEq>::eq` at /git/libc/src/unix/bsd/freebsdlike/freebsd/mod.rs:1445:29
note: inside `main` at src/main.rs:12:15
    --> src/main.rs:12:15
     |
12   |     let _eq = sigevent == sigevent;
     |               ^^^^^^^^^^^^^^^^^^^^
     = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Reply:
I agree.  I think we should remove most trait impls for any struct that contains a union (see #2816).  But how to do it without breaking existing code?